File: LeetCode/474.Ones_and_Zeroes/474.Ones_and_Zeroes.py
#!/usr/bin/env python3


class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        # largest subset of strs at most m 0 and n 1

        # knapsack problem
        # dp[i][j] means largest subset that are at most i 0 and j 1
        # dp[i][j] = max(dp[i-x][j-y] + 1) for x, y in strs

        # convert strs to 1 and 0 tuples
        zero_one = []
        for e in strs:
            z, o = 0, 0
            for c in e:
                if c == "0":
                    z += 1
                elif c == "1":
                    o += 1
            zero_one.append((z, o))

        # print(zero_one)

        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for x, y in zero_one:
            for i in range(m, x - 1, -1):
                for j in range(n, y - 1, -1):
                    dp[i][j] = max(dp[i][j], dp[i - x][j - y] + 1)

        # print(dp)

        return dp[m][n]

File: LeetCode/474.Ones_and_Zeroes/474.Ones_and_Zeroes.java
class Solution {
    public int findMaxForm(String[] strs, int m, int n) {
        int[][] dp = new int[m + 1][n + 1];

        for (int i = 0; i < strs.length; i++) {
            String str = strs[i];
            int ones = getTheNumOfOnes(str);
            int zeros = str.length() - ones;
            for (int j = m; j >= zeros; j--) {
                for (int k = n; k >= ones; k--) {
                    dp[j][k] = Math.max(dp[j][k], dp[j - zeros][k - ones] + 1);
                }
            }
        }

        return dp[m][n];
    }

    private int getTheNumOfOnes(String str) {
        char[] chars = str.toCharArray();
        int res = 0;
        for (char c : chars) {
            if (c == '1')
                res += 1;
        }
        return res;
    }
}